### DISCUSSION TITLE
Asking for help with sqlx prepare as a derivation

### DISCUSSION THREAD

**Discussion by ghost** [ipetkov/crane#704] [General]
Hey :)

I am trying to set up a rust project and mostly everything works. I followed the tutorial and API docs to get a decent setup. I am building with sqlx however, which mean adding a preBuild and postBuild and also for the checks as a commonArgs. I really wasn't a big fan of the prospect of always running the migrations on every iteration, so I thought I'd add a derivation that takes the cargo sqlx prepare output add that as a dependency for the crane cargo calls, that would be quite a bit more efficient.
I ran into the issue of cargo-sqlx wanting to check cargo metadata (it runs it as a `Command::new.args("").output()`). 
That wants to have network, but also fails when building with --no-sandbox. 
When I add the cargo metadata that fails complaining about not getting any network.
I tried inspecting why the cargo sqlx, cargo metadata command fails, but I didnt find anything that is very conclusive. So most likely the same thing that just calling cargo metadata runs into.

I have a reproduction at https://github.com/git-girl/min_repro_metadata
`nix run github:git-girl/min_repro_metadata`

It would be really kind if you know how I could set this up to give me some tips. 
 
I'm kinda out of ideas (except writing a custom sqlx command but I suffered last time I tried doing that for cargo watch and now I'm scared). 
Since in this repro I'm not overriding the scope i don't think it's related to #353.
I didn't see how this could be related to nix stripping file stat data. Cargo handles that okay otherwise as well, afaik. It seems quite connected to the metadata cargo invocation and I wanted to get around that by providing artifacts to it which didn't work. 
I also haven't checked a workaround like this one:
`cargo +nightly metadata -Z unstable-options -Zconfig-include --config config.toml --offline`
from https://github.com/nix-community/crate2nix/issues/148
(cargo metadata would have to then be smart enough not to try and make the request even if invoked without those flags again by sqlx though)

Best Regards

**Comment by ipetkov** [comment]
Hi @git-girl ! I think what you might be after is [`craneLib.mkCargoDerivation`](https://github.com/ipetkov/crane/blob/7e4586bad4e3f8f97a9271def747cf58c4b68f3c/docs/API.md#cranelibmkcargoderivation). Basically this function will basically handle source vendoring for you which will satisfy the `cargo metadata` calls (which are otherwise lazy and will try to hit the network if the sources aren't already available).

You can swap `pkgs.stdenv.mkDerivation` for `craneLib.mkCargoDerivation` in your flake and you'll be _almost_ there. `mkCargoDerivation` has a few additional required arguments, but if you are already defining your own build/check/install phases you can use some dummy values. Take a look at the docs for more context, and feel free to ask follow up questions if things don't make sense!

**Reply by ghost** [reply]
Oh, yes thank you! That makes a lot of sense. I think I had a setup that had the sqlx prepare as a mkCargoDerivation and iirc it did work similar to my current working setup, with just start postgres all the time (currently I just pass it as commonArgs, but I did start by having it on the derivation). But I didn't think to use this for sqlx prepare. I think maybe the issue I'd have there again is getting rid of this behavior for some things, where I don't wan't this (cargoFmt) but also reuuse the craneLib instance, because caching. I'll try doing some cursed variable thing for the cargoDerivation. Or just accept that I'll need 2 craneLib instances? I will check it out in the next few days and will report back :)
Thanks!
